networkx/algorithms/centrality/domirank.py

```diff
@@ -0,0 +1,298 @@
+
+    dt: float, optional (default=0.1)
+     The step size for the Newton iteration.
+
+    epsilon: float, optional (default=1e-5)
+     The relative stopping criterion for convergence.
+
+    max_iter: integer, optional (default=50)
+      Maximum number of Newton iterations allowed.
+
+    check_step: integer, optional (default=10)
+     The number of steps between checking whether or
+     not DomiRank has converged/diverged.
+     Note, this paramter must satisfy check_step > 2.
+
+
+    Returns
+    -------
+    nodes : dictionary
+       Dictionary of nodes with DomiRank centrality as the value.
+
+    sigma : float
+       The input $\sigma \in [0,1]$ normalized by the smallest eigenvalue
+
+    converged: boolean
+       A boolean representing whether or not the method did not diverge.
+
+    Examples
+    --------
+    >>> G = nx.path_graph(5)
+    >>> centrality, sigma, converged = nx.domirank(G)
+    >>> print([f"{node} {centrality[node]:0.2f}" for node in centrality])
+    ['0 -0.54', '1 1.98', '2 -1.08', '3 1.98', '4 -0.54']
+
+    Raises
+    ------
+    NetworkXPointlessConcept
+        If the graph G is the null graph.
+
+    NetworkXError
+        If the input for G is not a network.graph.classes.Graph object
+
+    Warning
+        If one is supercharging the competition parameter for the
+        analytical solution.
+    Warning
+        If one is using the analytical solution for a large system
+        - i.e. len(G) > 5000, as the algorithm will be slow.
+
+    See Also
+    --------
+    :func:`scipy.sparse.linalg.solve`
+    :func:`~networkx.algorithms.link_analysis.pagerank_alg.pagerank`
+    :func:`~networkx.algorithms.link_analysis.hits_alg.hits`
+
+    References
+    ----------
+    .. [1] Engsig, M., Tejedor, A., Moreno, Y. et al.
+    "DomiRank Centrality reveals structural fragility of complex networks via node dominance."
+    Nat Commun 15, 56 (2024). https://doi.org/10.1038/s41467-023-44257-0
+    """
+    import numpy as np
+    import scipy as sp
+
+    if (
+        type(G) == nx.classes.graph.Graph or type(G) == nx.classes.digraph.DiGraph
+    ):  # check if it is a networkx Graph
+        if len(G) == 0:
+            raise nx.NetworkXPointlessConcept(
+                "cannot compute centrality for the null graph"
+            )
+        GAdj = nx.to_scipy_sparse_array(G)  # convert to scipy sparse csr array
+    else:
+        raise nx.NetworkXError(
+            "can only compute the centrality for nx.classes.graph.Graph object"
+        )
+    # Here we renormalize sigma with the smallest eigenvalue (most negative eigenvalue) by calling the "hidden" function _find_smallest_eigenvalue()
+    # Note, this function always uses the recursive definition
+    if analytical == False:
+        # If the recursive formulation is used, the sigma has to be bounded (competition parameter).
+        if sigma < 0 or sigma > 1:
+            raise nx.NetworkXPointlessConcept(
```

Comment:
From what I gather, `NetworkXPointlessConcept` refers specifically to the null graph; it's not supposed to be a catch-all for "meaningless" parameter values.

Reply:
I don't personally feel qualified enough to be the arbiter of what's good, but it makes sense to me. I'm sure @dschult or @rossbar could give more complete answers to your questions.

For what it's worth, I think the error message given when `analytical=False` should also mention that `sigma > 1` is only allowed when `analytical=True`; people using `sigma > 1` with `analytical=False` are likely to have just put the wrong value for `analytical`.

Reply:
Thank you, this is indeed a great point. I will make the amendment now.